src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -379,6 +461,63 @@ case class FilterRows(
   }
 }
 
+case class FilterRowsIR(
+  child: MatrixIR,
+  pred: IR) extends MatrixIR {
+
+  def children: IndexedSeq[BaseIR] = Array(child, pred)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): FilterRowsIR = {
+    assert(newChildren.length == 2)
+    FilterRowsIR(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
+  }
+
+  def typ: MatrixType = child.typ
+
+  def execute(hc: HailContext): MatrixValue = {
+    val prev = child.execute(hc)
+
+    val ec = prev.typ.rowEC
+
+    val (rTyp, predCompiledFunc) = ir.Compile[Long, Long, Boolean](
+      "va",     typ.rvRowType,
+      "global", typ.globalType,
+      pred
+    )
+
+    val aggregatorOption = Aggregators.buildRowAggregations(
+      prev.rvd.sparkContext, prev.typ, prev.globals, prev.colValues, ec)
+    //
+    // Everything used inside the Spark iteration must be serializable,
+    // so we pick out precisely what is needed.
+    //
+    val fullRowType = prev.typ.rvRowType
+    val localRowType = prev.typ.rowType
+    val localEntriesIndex = prev.typ.entriesIdx
+    val localGlobalType = typ.globalType
+    val localGlobals = prev.globals.broadcast
+
+    val filteredRDD = prev.rvd.mapPartitionsPreservesPartitioning(prev.typ.orvdType) { it =>
+      val fullRow = new UnsafeRow(fullRowType)
+      val row = fullRow.deleteField(localEntriesIndex)
+      it.filter { rv =>
+        fullRow.set(rv)
+        ec.set(0, localGlobals.value)
+        ec.set(1, row)
+        aggregatorOption.foreach(_ (rv))
```

Comment:
Same comment about aggregations, which means we can get rid of all the UnsafeRow stuff.

Reply:
done